tools/calyx-ffi-macro/src/lib.rs

```diff
@@ -0,0 +1,404 @@
+            .expect("caller of calyx_ffi did not use cargo to build project")
+            .1,
+    );
+
+    let args = parse_macro_input!(attrs as CalyxFFIMacroArgs);
+    let item_struct = parse_macro_input!(item as syn::ItemStruct);
+    let name = item_struct.ident;
+    let given_path = args.src.to_string_lossy().to_string();
+
+    let mut path = source_manifest_dir;
+    path.push(given_path);
+    let path = path;
+
+    // <sorry>
+    let comp = calyx::parse_calyx_file(&args, path.clone());
+    if let Err(error) = comp {
+        return error;
+    }
+    let comp = comp.unwrap();
+    let comp = comp.get();
+    // </sorry>
+
+    let comp_name =
+        syn::parse_str::<syn::LitStr>(&format!("\"{}\"", comp.name))
+            .expect("failed to turn quoted name into string");
+    let comp_path = syn::parse_str::<syn::LitStr>(&format!(
+        "\"{}\"",
+        path.to_string_lossy()
+    ))
+    .expect("failed to turn quoted path into string");
+
+    let backend_macro = args.backend;
+    let mut input_names = Vec::new();
+    let mut output_names = Vec::new();
+    let mut fields = vec![];
+    let mut default_field_inits = vec![];
+    let mut getters = vec![];
+    let mut setters = vec![];
+    let mut width_getters = vec![];
+    let mut port_names = vec![];
+
+    for port in comp.signature.borrow().ports() {
+        let port_name_str = port.borrow().name.to_string();
+        let port_name = syn::parse_str::<syn::Ident>(&port_name_str)
+            .expect("failed to turn port name into identifier");
+
+        port_names.push(port_name.clone());
+
+        let port_width = port.borrow().width;
+        let width_getter = format_ident!("{}_width", port_name);
+        width_getters.push(quote! {
+            pub const fn #width_getter() -> u64 {
+                #port_width
+            }
+        });
+
+        default_field_inits.push(quote! {
+            #port_name: calyx_ffi::value_from_u64::<#port_width>(0)
+        });
+
+        // idk why input output ports are being flipped??
+        match port.borrow().direction.reverse() {
+            calyx_ir::Direction::Input => {
+                let setter = format_ident!("set_{}", port_name);
+                fields.push(quote! {
+                    pub #port_name: calyx_ffi::Value<#port_width>
+                });
+                setters.push(quote! {
+                    pub fn #setter(&mut self, value: u64) {
+                        self.#port_name = calyx_ffi::value_from_u64::<#port_width>(value);
+                    }
+                });
+                input_names.push(port_name);
+            }
+            calyx_ir::Direction::Output => {
+                fields.push(quote! {
+                    #port_name: calyx_ffi::Value<#port_width>
+
+                });
+
+                let bitvec_getter = format_ident!("{}_bits", port_name);
+
+                getters.push(quote! {
+                    pub fn #port_name(&self) -> u64 {
```

Comment:
It might make sense to skip generating these getters when the port is wider than 64 bits. Then this function would never fail at runtime.

Reply:
Good idea!